File: frontend/types/analytics.ts
export interface WeeklyEnrollmentPoint {
  week_start: string;
  enrollments: number;
}

export interface TopCourseItem {
  course_id: string;
  title: string;
  enrollments: number;
}

export interface RecentActiveUserItem {
  user_id: string;
  full_name: string;
  email: string;
  role: string;
  last_activity_at: string;
}

export interface StaleDraftCourseAlert {
  course_id: string;
  title: string;
  days_in_draft: number;
}

export interface AdminAnalytics {
  total_users: number;
  active_courses: number;
  enrollments_this_month: number;
  completion_rate: number;
  enrollments_by_week: WeeklyEnrollmentPoint[];
  top_courses: TopCourseItem[];
  recent_active_users: RecentActiveUserItem[];
  stale_draft_alerts: StaleDraftCourseAlert[];
}

export interface InstructorCourseStats {
  total: number;
  published: number;
  drafts: number;
}

export interface LowestQuizPerformance {
  quiz_id: string;
  quiz_title: string;
  course_id: string;
  course_title: string;
  average_score: number;
}

export interface InstructorActivityItem {
  type: "ENROLLMENT" | "QUIZ_ATTEMPT" | "LESSON_COMPLETED";
  timestamp: string;
  course_id: string;
  course_title: string;
  message: string;
}

export interface InstructorAnalytics {
  my_courses: InstructorCourseStats;
  total_enrolled_students: number;
  average_completion_rate: number;
  lowest_performing_quiz: LowestQuizPerformance | null;
  recent_activity: InstructorActivityItem[];
}

export interface StudentInProgressCourse {
  course_id: string;
  course_title: string;
  progress_pct: number;
  next_lesson_id: string | null;
  next_lesson_title: string | null;
}

export interface StudentCompletedCourse {
  course_id: string;
  course_title: string;
  completed_at: string;
}

export interface PendingQuizItem {
  quiz_id: string;
  quiz_title: string;
  course_id: string;
  course_title: string;
}

export interface StudentPersonalStats {
  hours_learned: number;
  streak_days: number;
  quiz_average: number;
}

export interface StudentLearningPathItem {
  path_id: string;
  title: string;
  progress_pct: number;
}

export interface StudentAnalytics {
  in_progress_courses: StudentInProgressCourse[];
  completed_courses: StudentCompletedCourse[];
  pending_quizzes: PendingQuizItem[];
  personal_stats: StudentPersonalStats;
  assigned_learning_paths: StudentLearningPathItem[];
}

File: backend/app/schemas/analytics.py
from datetime import date, datetime
from pydantic import BaseModel


class WeeklyEnrollmentPoint(BaseModel):
    week_start: date
    enrollments: int


class TopCourseItem(BaseModel):
    course_id: str
    title: str
    enrollments: int


class RecentActiveUserItem(BaseModel):
    user_id: str
    full_name: str
    email: str
    role: str
    last_activity_at: datetime


class StaleDraftCourseAlert(BaseModel):
    course_id: str
    title: str
    days_in_draft: int


class AdminAnalyticsOut(BaseModel):
    total_users: int
    active_courses: int
    enrollments_this_month: int
    completion_rate: float
    enrollments_by_week: list[WeeklyEnrollmentPoint]
    top_courses: list[TopCourseItem]
    recent_active_users: list[RecentActiveUserItem]
    stale_draft_alerts: list[StaleDraftCourseAlert]


class InstructorCourseStats(BaseModel):
    total: int
    published: int
    drafts: int


class LowestQuizPerformance(BaseModel):
    quiz_id: str
    quiz_title: str
    course_id: str
    course_title: str
    average_score: float


class InstructorActivityItem(BaseModel):
    type: str
    timestamp: datetime
    course_id: str
    course_title: str
    message: str


class InstructorAnalyticsOut(BaseModel):
    my_courses: InstructorCourseStats
    total_enrolled_students: int
    average_completion_rate: float
    lowest_performing_quiz: LowestQuizPerformance | None
    recent_activity: list[InstructorActivityItem]


class StudentInProgressCourse(BaseModel):
    course_id: str
    course_title: str
    progress_pct: float
    next_lesson_id: str | None
    next_lesson_title: str | None


class StudentCompletedCourse(BaseModel):
    course_id: str
    course_title: str
    completed_at: datetime


class PendingQuizItem(BaseModel):
    quiz_id: str
    quiz_title: str
    course_id: str
    course_title: str


class StudentPersonalStats(BaseModel):
    hours_learned: float
    streak_days: int
    quiz_average: float


class StudentLearningPathItem(BaseModel):
    path_id: str
    title: str
    progress_pct: float


class StudentAnalyticsOut(BaseModel):
    in_progress_courses: list[StudentInProgressCourse]
    completed_courses: list[StudentCompletedCourse]
    pending_quizzes: list[PendingQuizItem]
    personal_stats: StudentPersonalStats
    assigned_learning_paths: list[StudentLearningPathItem]
